main/src/ca/uwaterloo/flix/language/phase/TreeShaker.scala

```diff
@@ -18,29 +18,210 @@ package ca.uwaterloo.flix.language.phase
+    }
+
+    /**
+      * Returns the function symbols reachable from `hs`.
+      */
+    def visitHeadTerms(hs: List[SimplifiedAst.Term.Head]): Set[Symbol.DefnSym] = hs.map(visitHeadTerm).fold(Set())(_++_)
+
+    /**
+      * Returns the function symbols reachable from the given SimplifiedAst.Term.Head `head`.
+      */
+    def visitHeadTerm(h0: SimplifiedAst.Term.Head): Set[Symbol.DefnSym] = {
+      h0 match {
+        case SimplifiedAst.Term.Head.Lit(lit, tpe, loc) => visitExp(lit)
+        case SimplifiedAst.Term.Head.App(sym, args, tpe, loc) => Set(sym)
+        case _ => Set.empty
+      }
+    }
+
+    /**
+      * Returns the function symbols reachable from `bs`.
+      */
+    def visitBodyTerms(bs: List[SimplifiedAst.Term.Body]): Set[Symbol.DefnSym] = bs.map(visitBodyTerm).fold(Set())(_++_)
+
+    /**
+      * Returns the function symbols reachable from the given SimplifiedAst.Term.Body `body`.
+      */
+    def visitBodyTerm(b0: SimplifiedAst.Term.Body): Set[Symbol.DefnSym] = {
+      b0 match {
+        case SimplifiedAst.Term.Body.Lit(exp, tpe, loc) => visitExp(exp)
+        case _ => Set.empty
+      }
+    }
+
+    /**
+      * Returns the function symbols reachable from `es`.
+      */
+    def visitExps(es: List[Expression]): Set[Symbol.DefnSym] =  es.map(visitExp).fold(Set())(_++_)
+
+    /**
+      * Returns the function symbols reachable from the given Expression `e0`.
+      */
+    def visitExp(e0: Expression): Set[Symbol.DefnSym] =  e0 match {
+      case Expression.Ref(sym, tpe, loc) => Set(sym)
+      case Expression.Lambda(args, body, tpe, loc) => visitExp(body)
+      case Expression.Hook(hook, tpe, loc) => Set(hook.sym)
+      case Expression.MkClosure(lambda, freeVars, tpe, loc) => visitExp(lambda)
+      case Expression.MkClosureRef(ref, freeVars, tpe, loc) => visitExp(ref)
+      case Expression.ApplyRef(sym, args, tpe, loc) => visitExps(args) + sym
+      case Expression.ApplyTail(sym, formals, actuals, tpe, loc) => visitExps(actuals) + sym
+      case Expression.Apply(exp, args, tpe, loc) => visitExps(args) ++ visitExp(exp)
+      case Expression.Unary(op, exp, tpe, loc) => visitExp(exp)
+      case Expression.Binary(op, exp1, exp2, tpe, loc) => visitExp(exp1) ++ visitExp(exp2)
+      case Expression.IfThenElse(exp1, exp2, exp3, tpe, loc) => visitExp(exp1) ++ visitExp(exp2) ++ visitExp(exp3)
+      case Expression.Let(sym, exp1, exp2, tpe, loc) => visitExp(exp1) ++ visitExp(exp2)
+      case Expression.Is(exp, tag, loc) => visitExp(exp)
+      case Expression.Tag(sym, tag, exp, tpe, loc) => visitExp(exp)
+      case Expression.Untag(tag, exp, tpe, loc) => visitExp(exp)
+      case Expression.Index(base, offset, tpe, loc) => visitExp(base)
+      case Expression.Tuple(elms, tpe, loc) => visitExps(elms)
+      case Expression.Existential(fparam, exp, loc) => visitExp(exp)
+      case Expression.Universal(fparam, exp, loc) => visitExp(exp)
+      case _ => Set.empty
```

Comment:
We try to write out all cases, even if they are trivial. For example, `Int8 => ...`. The reason is that when we add a new AST node we immediately get warnings from where we have not handled it. If we use a default case `_` then it will catch any newly added AST nodes which is probably not what we want.